### DISCUSSION TITLE
Import-Module: Could not load file or assembly

### DISCUSSION THREAD

**Discussion by MattK-CPA** [pnp/powershell#58] [General]
**Import-Module PnP.PowerShell**
Import-Module: Could not load file or assembly 'Microsoft.Online.SharePoint.Client.Tenant, Version=16.1.0.0, Culture=neutral, PublicKeyToken=71e9bce111e9429c'. Could not find or load a specific file. (0x80131621)

Hello I recently installed PowerShell 7.1.0 and then installed the recent version of PnP.PowerShell

Name                           Value
PSVersion                      7.1.0
PSEdition                      Core
GitCommitId                    7.1.0
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

Install-Module -Name PnP.PowerShell -AllowPrerelease

However, when I attempt to import the module for PnP.PowerShell it throws errors.  
Is anyone willing to give me some hints or tips about what I have done wrong or what I have not done that I need to do to be able to get PnP.PowerShell to successfully work.

Thanks.

**Comment by erwinvanhunen** [comment]
You might have currently another module loaded that uses a different version of this assembly. 

To check if a version is currently loaded in PowerShell:

```powershell
[System.AppDomain]::CurrentDomain.GetAssemblies() | Where-Object{$_.Location -like "*Microsoft.Online.SharePoint.Client.Tenant*"}
```

Both Microsoft.Online.SharePoint.PowerShell (e.g. SharePoint Online Management Shell) and PnP PowerShell (both the legacy and the current version) refere to this assembly? Maybe you still have the older version of PnP PowerShell installed?

To update the SharePoint Online Management Shell to the latest version:

```powershell
Update-Module -Name Microsoft.Online.SharePoint.PowerShell
```

To uninstall the legacy version of PnP PowerShell:

```powershell
Uninstall-Module -Name SharePointPnPPowerShellOnline -AllVersions -Force
```

**Reply by MattK-CPA** [reply]
Thanks @erwinvanhunen.

The Where-Object query that you provided does return a result: 
Location: C:\Users\Username\Documents\PowerShell\Modules\Microsoft.Online.SharePoint.PowerShell\16.0.20717.12000\Microsoft.Online.SharePoint.Client.Tenant.dll
but this is essentially a recent version of SharePoint Online Management Shell. 

I admit that it was a surprise to find that I could do an Install-Module, Update-Module to get Microsoft.Online.SharePoint.PowerShell from a repository.  I missed the communications on that being available.  So, Thanks.  I had still being using a Microsoft download MSI package of the SharePoint Online Management Shell to make the module available (especially in the PowerShell Desktop Version 5.1).  So that is nice to know. 

When I do a  get-Module  there is **no** SharePointPnPPowerShellOnline module in the results.  
But, when I do  get-Module -ListAvailable  there is a SharePointPnPPowerShellOnline 3.28.2012.0 showing in the Directory: C:\Program Files\WindowsPowerShell\Modules
I don't remember actually installing it from within PowerShell 7 and the directory suggests that it is from a previous Windows Powershell version.  Is that still "seen" in the pwsh7 environment and by the PnP.PowerShell module ?

**Comment by erwinvanhunen** [comment]
We're a bit puzzled about the version included with Microsoft.SharePointOnline.PowerShell. That version is actually not targeting SharePoint Online, it's the build for SP 2016. We are including version 16.1.* in our builds, which is always the latest. I just tried on my machine by opening PS5, then loading Connect-SPOService (which will load the Microsoft.Online.SharePoint.Client.Tenant.dll assembly into the current appdomain). Then I used Connect-PnPOnline and received no error message. When checking the appdomain I noticed that both versions of the Microsoft.Online.SharePoint.Client.Tenant assemblies where loaded.

Can you uninstall SharePointPnPPowerShellOnline and try again?

**Reply by erwinvanhunen** [reply]
Update: we figured out the version number difference of the versions used by Microsoft.SharePointOnline.PowerShell. It's by design. Long story :-)

**Reply by MattK-CPA** [reply]
Hello
I did attempt the uninstall of SharePointPnPPowerShellOnline from inside PowerShell 7.
`Uninstall-Module -Name SharePointPnPPowerShellOnline -AllVersions -Force`
Message produced: No match was found for the specified search criteria and module names 'SharePointPnPPowerShellOnline'.
When I then do
`get-Module -ListAvailable`
it still showed SharePointPnPPowerShellOnline 3.28.2012.0 in Directory: C:\Program Files\WindowsPowerShell\Modules
That was the reason I asked - Is that still "seen" in the pwsh7 environment and by the PnP.PowerShell module ?

Anyway,  I opened up a Windows PowerShell 5 console and I did the 
`Uninstall-Module -Name SharePointPnPPowerShellOnline -AllVersions -Force`
there also.  I then did a re-install of the module in  Windows PowerShell 5 ~ because I needed to do some actions against a site. Exit the console.

When I went and re-opened the PowerShell 7 console, it still shows SharePointPnPPowerShellOnline 3.28.2012.0 as an available module.
Try again, with the
`Import-Module PnP.PowerShell`
Loaded. No errors. Can connect and so on using the cmdlets. Good ! ? !

Thanks again @erwinvanhunen for being willing to engage and offer advice.

**Reply by erwinvanhunen** [reply]
Check the environment variable PSModulePath in PowerShell 7 and in PowerShell 5:

```powershell
$env:PSModulePath
```

All the paths in there are being resolved by PowerShell when it is looking for modules. Most likely SharePointPnPPowerShellOnline is installed in one of those paths. 

Most likely both PowerShell 7 and PowerShell 5 have references to the same folder where the version 3.28.2012.0 is installed.

The `Import-Module -Name PnP.PowerShell` you did is absolutely the way to do it when going forward, until you're ready to fully uninstall 3.28.2012.0 (notice that PnP.PowerShell is in principle a rewritten copy of version 3.28.2012.0 and most of the code has not changed, so very likely you will be able to use PnP.PowerShell for your tasks instead of SharePointPnPPowerShellOnline).

If you launch PowerShell and without doing an `import-module` first enter `connect-pnponline` (or another other PnP PowerShell cmdlet) then PowerShell will try to automatically load the first match it finds when following the paths in PSModulePath environment variable.

**Reply by MattK-CPA** [reply]
So ... it seems that you are suggesting that it should be possible to run Module **SharePointPnPPowerShellOnline** (pnp/pnp-powershell) in **PowerShell5** and module **PnP.PowerShell** (pnp/powershell) in **PowerShell7** side-by-side on the same Windows OS.  But for this to work satisfactorily, that an implicit Import-module must be executed to ensure that the appropriate module is loaded (preventing accidental Cmdlet matching confusion to the wrong module).
But the first instruction about Upgrading (https://pnp.github.io/powershell/articles/upgrading.html) is to "make sure to uninstall any edition of PnP PowerShell that runs on the Windows PowerShell version (the one with the blue background)".
I do not mean to be difficult - I am just trying to understand whether it is an all or nothing situation.  It would be nice to think that I could continue to use PnP in PowerShell5 while I get the PnP in PowerShell7 working satisfactorily (I continue to experience problems with PnP in PowerShell7 - maybe it has something to do with MFA requirements).
_Thank you_.

**Reply by ToddKlindt** [reply]
I am currently running both modules in both PowerShells. I outlined how I'm doing it in this blog post, https://www.toddklindt.com/blog/Lists/Posts/Post.aspx?ID=882 If you want the tl;dr, jump to the "Using Both Modules in Both PowerShells" section.

**Comment by maxferrario-msf** [comment]
Hi, I've just followed the instructions on upgrading from the Legacy version of PnP PowerShell (https://pnp.github.io/powershell/articles/upgrading.html) and - after uninstalling SharePointPnPPowerShellOnline and installing PnP.PowerShell - had a similar issue:

> ```
> PS> Connect-PnPOnline
> Connect-PnPOnline : The 'Connect-PnPOnline' command was found in the module 'PnP.PowerShell', but the module could not be loaded. For more information, run 'Import-Module PnP.PowerShell'.
> PS> Import-Module PnP.PowerShell
> Import-Module : Could not load file or assembly 'System.Memory, Version=4.0.1.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.
> ```

I only have windows powershell installed on my windows 10 system (PSVersion = 5.1.19041.610).
`Get-Module -ListAvailable` doesn't show any trace of the old SharePointPnPPowerShellOnline, and `[System.AppDomain]::CurrentDomain.GetAssemblies() | Where-Object{$_.Location -like "*Microsoft.Online.SharePoint.Client.Tenant*"}` only returns `D:\users\omissis\documents\WindowsPowerShell\Modules\PnP.PowerShell\0.3.19\Framework\Microsoft.Online.SharePoint.Client.Tenant.dll`

Any idea on how to solve this?

**Reply by maxferrario-msf** [reply]
[Please tell me if you feel it's better to open a new discussion about this specific issue]

**Reply by maxferrario-msf** [reply]
If I type

> `PS > Add-Type -Path 'd:\users\omissis\documents\WindowsPowerShell\modules\PnP.PowerShell\0.3.19\Common\System.Memory.dll'`

before importing the module everything goes fine.
Now I _only_ need to find a way to do this in my script... even if this is a very bad workaround, I will at least be able to test my script aginst pnp.powershell!

**Reply by erwinvanhunen** [reply]
We found an issue with system.memory.dll which has been resolved in later releases. Please update your module install and it should be fixed.

**Reply by maxferrario-msf** [reply]
Thanks Erwin
